### ISSUE TITLE
systemd 233 breaks support for notify services using PAM

### ISSUE THREAD

**Issue by sre** [systemd/systemd#6045]
### Submission type

 - Bug report

### systemd version the issue has been seen with

> version 233

### Used distribution

> yocto based

### Bug Report

> systemd 233 does no longer support services, which are `Type=notify` and use PAM. I provide a simple testcase below (of course the testcase should go into failed state after 10 seconds due to missing READY, but it goes directly into failed state due to empty cgroup).

```
[Unit]
Description=Testcase for Type=notify + PAM

[Service]
Type=notify
ExecStart=/bin/sleep 10
User=nobody
PAMName=login
```

### Details

If a service in systemd uses PAM, a new session is created using systemd-logind. This removes the service's PID from the cgroup systemd created for it. That triggers `service_notify_cgroup_empty_event()` for the service. I suspect, that this has not been noticed so far, since "cgroup is empty" is only printed with systemd in debug mode and the event was mostly ignored until v233. Since v233 `service_notify_cgroup_empty_event()` instantly kills services with Type=notify, though (https://github.com/systemd/systemd/blob/v233/src/core/service.c#L2593).

**Comment by sre** [comment]
I tested, that the regression can be fixed by removing the following code block from `service_notify_cgroup_empty_event()`, that was added in v233:

```
if (s->type == SERVICE_NOTIFY) {
        /* No chance of getting a ready notification anymore */
        service_enter_stop_post(s, SERVICE_FAILURE_PROTOCOL);
        break;
}
```

I can send a patch for that, but I assume, that the actual bug is systemd-logind removing the process from the cgroup systemd created.

**Comment by joukewitteveen** [comment]
A similar logic is applied to pidfiles:
```
if (s->pid_file_pathspec) {
        /* Give up hoping for the daemon to write its PID file */
        log_unit_warning(u, "Daemon never wrote its PID file. Failing.");

        service_unwatch_pid_file(s);
        if (s->state == SERVICE_START)
                service_enter_stop_post(s, SERVICE_FAILURE_PROTOCOL);
        else
                service_enter_stop(s, SERVICE_FAILURE_PROTOCOL);
}
break;
```

Can you test the testcase below?
```
[Unit]
Description=Testcase for Type=forking + PAM

[Service]
Type=forking
PIDFile=/tmp/na.pid
ExecStart=/bin/sleep 10
User=nobody
PAMName=login
```
It works fine (fails only after 10 seconds) on my systemd 232 system. I currently do not have a 233 system at hand. I expect this unit to also go into a failed state directly on your system because of an empty cgroup.

I do not think this bug is related to notify services, but rather to cgroup management.

**Comment by sre** [comment]
@joukewitteveen systemd v233 is now available in Debian testing. Your testcase seems to work ok (only fails after 10 seconds) on v233. The testcase I provided also fails instantly on Debian's systemd v233.

**Comment by joukewitteveen** [comment]
Indeed, judging by the journal it looks like the cgroup_empty event is not triggered in the `forking + PAM` testcase! So why is it triggered in the `notify + PAM` testcase?

**Comment by joukewitteveen** [comment]
On closer inspection, the cgroup_empty event appears to be triggered for the `forking + PAM` testcase too, but apparently `pid_file_pathspec` is not set (yet?) so the service is not terminated.

I suspect this extra cgroup_epmty event is caused by the extra process forked in setting up PAM. In fact, there are three places where we `fork()` in `src/core/execute.c`:
- `setup_pam`
- `setup_private_users`
- `exec_spawn`

I guess the current logic implemented for the cgroup_empty event does not take processes initiated by the first two into account. Maybe the presence of these processes can even prevent proper cgroup_empty events from being triggered.

If this is indeed the cause of this bug, we might want to find a way to exclude these control processes from our cgroup watch. This might be difficult since the only cgroup notifications we get are for going from empty to non-empty and from non-empty to empty.

**Comment by joukewitteveen** [comment]
Here is a better testcase, which should maybe be added to the regression tests.

    [Unit]
    Description=Testcase for Type=notify + PAM
    
    [Service]
    Type=notify
    RemainAfterExit=yes
    ExecStart=/usr/bin/systemd-notify --ready
    User=nobody
    PAMName=login

Without the last line (`PAMName=login`), the service works fine. With the line included, current systemd receives a cgroup_empty event before the main process is started and kills the process immediately.

~~The bug can be resolved by moving the [initiation of the cgroup watch](https://github.com/systemd/systemd/blob/v234/src/core/cgroup.c#L1330-L1331) either to [the execution stage](https://github.com/systemd/systemd/blob/v234/src/core/execute.c#L3040) or even the [service management](https://github.com/systemd/systemd/blob/v234/src/core/service.c#L1388). However, this does mean that `unit_watch_cgroup` must be called explicitly for every other invocation of `unit_realize_cgroup_now` as well.~~ Alternatively, we could try to ignore all cgroup_empty events that occur [while/before service_spawn is running](https://github.com/systemd/systemd/blob/v234/src/core/service.c#L1243-L1381), but that may be a lot more tricky.

Edit: I tried moving `unit_watch_cgroup` to the execution stage, but it does not get rid of the unwanted cgroup_empty event.

**Comment by joukewitteveen** [comment]
I believe I have found the cause of this issue.
The problem is that we add the cgroup watch to the `system.slice`, whereas we move (using [`unit_attach_pids_to_cgroup`](https://github.com/systemd/systemd/blob/v234/src/core/scope.c#L338)) the processes to the `user-XX.slice` when `PamName=` is set. Hence we receive the wrong cgroup_empty notifications (and do not receive the right ones).

**Comment by poettering** [comment]
Proposed fix waiting in #6932.
